src/garage/envs/base.py

```diff
@@ -140,6 +140,11 @@ def step(self, action):
         """
         return self.env.step(action)
 
+    def __getstate__(self):
+        """See `Object.__getstate__."""
+        self.close()
```

Comment:
I'm not convinced this is the right thing to do. Pickling an object shouldn't modify its state.

Also, most environments already know how to pickle themselves, so it's probably best to just rely on that (and not define `__getstate__` and `__setstate__` for this class).

Is there some environment in particular that needs this?

Reply:
after a second thought, actually it meant to be closing its viewer

Reply:
Hmm yeah why do we need to close the environment to pickle it? The original env should be able to live on.

Reply:
This is meant to be closing the viewer

Reply:
It's not valid to call *any* method on `gym.Env` after calling `gym.Env.close()`. That means your environment is inoperable after pickling.

Also, why should the viewer close when I pickle it? What if I was using that viewer?

Reply:
To make it clear, instead of calling `self.close()`, we call `self._close_viewer_window`
(https://github.com/rlworkgroup/garage/blob/7b6d7a528f623d8e6c985cf7ee4d39c4b73fdebd/src/garage/envs/base.py#L70).

If a viewer object is not available when needed, it will create one, e.g. https://github.com/openai/gym/blob/master/gym/envs/box2d/car_racing.py#L345

Reply:
we can't close the viewer while we're pickling. you can make a copy of `self.__dict__` and delete the viewer reference from that copy, but you can't close the viewer.

imagine i'm running a training loop like the one below. the viewer window would flicker on the screen.
```python
env = gym.make('Foo-v1')
env.render()  # opens viewer window
algo = FooAlgo()

while training:
    paths = obtain_samples(policy, env, batch_size)
    algo.train_once(paths)
    save_snapshot(env, policy)  # closes viewer window
    rollout(env, policy, max_path_length=100, render=True)  # opens viewer window
```

Reply:
emm ok. It makes sense. I will remove it from `self.__dict__`.

Reply:
turns out we can't remove the viewer object, gym.Env assumes it's always there. If we remove that, we will get
```
    if self.viewer is None:
AttributeError: 'CartPoleEnv' object has no attribute 'viewer'
```

Reply:
can you remove it from a copy of self.__dict__?

Reply:
At the same time we pickle and unpickle the envs and use it
https://github.com/rlworkgroup/garage/blob/master/src/garage/sampler/on_policy_vectorized_sampler.py#L29
so we can't really remove the viewer attribute, but at the same time we can't have it being a `viewer` object.

Reply:
How about
```python
    def __getstate__(self):
        """See `Object.__getstate__."""
        _state = self.__dict__.copy()
        _state['env'].__dict__['env'].__dict__['viewer'] = None
        return _state

    def __setstate__(self, state):
        """See `Object.__setstate__.

        Args:
            state (dict): Unpickled state of this object.

        """
        obj = type(self)(state['_env'], state['_env_name'])
        self.__dict__.update(obj.__dict__)
```

Reply:
You need to do something like this:
```python
state = self.__dict__.copy()
if 'viewer' in state:
    state['viewer'] = None
return state
```

Reply:
yes, just the viewer object is deep down, inside the nested env

Reply:
You probably need something like this:
```python
    def __getstate__(self):
        """See `Object.__getstate__."""
        state = self.__dict__.copy()
        state['env'] = copy.copy(state['env'])
        if 'viewer' in state['env'].__dict__:
            state['env'].__dict__ = None
        return state

    def __setstate__(self, state):
        """See `Object.__setstate__.

        Args:
            state (dict): Unpickled state of this object.

        """
        self.__dict__ = state
```

Reply:
why do we need to copy `state['env']`? If we already have a copy of `self.__dict__`, shouldn't we just assign the viewer to `None`? like
```python
    def __getstate__(self):
        """See `Object.__getstate__."""
        state = self.__dict__.copy()
        # state['env'] = copy.copy(state['env'])  
        if 'viewer' in state['env'].__dict__['env'].__dict__:
            state['env'].__dict__['env'].__dict__['viewer'] = None
        return state
```
And we don't even need `__setstate__` anymore

Reply:
The copy of `self.__dict__` is shallow, so `state['env']` still refers to the original environment. If you modify that, you're modifying the original environment.

Reply:
How about we change the first line to
state = self.__dict__.deepcopy()

Reply:
Turns out we can't deep copy the dict if a viewer instance is in there. Here's my solution
```python
    def __getstate__(self):
        # we first make a copy of the viewer
        _viewer = self.__dict__['env'].__dict__['env'].__dict__['viewer']
        # remove the viewer and make a copy of the state
        self.__dict__['env'].__dict__['env'].__dict__['viewer'] = None
        state = copy.deepcopy(self.__dict__)
        # assign the viewer back to self.__dict__
        self.__dict__['env'].__dict__['env'].__dict__['viewer'] = _viewer
        # the returned state doesn't have the viewer
        return state
```

Reply:
i don't think this *exact* solution can work in general, because we have no idea how many wrappers are around `env` there are.

You could detect a `gym.Wrapper` and use `Wrapper.unwrapped` though.

Which gym envs have this problem? is it only the MuJoCo ones?

Reply:
also, do you have any insight into why this passes https://github.com/rlworkgroup/garage/blob/master/tests/garage/tf/envs/test_base.py?

Reply:
Yeah, to begin with all environments should be pickleable without need for a wrapper.

If this continues to be a problem, perhaps we should use constructor objects after all. Alternatively, we could monkey patch the viewer type to be "pickleable" by throwing away its state.

Reply:
> https://github.com/rlworkgroup/garage/blob/master/tests/garage/tf/envs/test_base.py

I got segmentation fault when running the test.